### PR TITLE
Added Youtube embed player

### PR DESCRIPTION
**Pull request by ayes-web**
Type: [enhancement]
Notes: Adds Youtube video embed player
Fixes: vector-im/element-web/issues/3466

Before: 
before

After:
after
after2

---
This PR currently has no changelog labels, so will not be included in changelogs.

A reviewer can add one of: `T-Deprecation`, `T-Enhancement`, `T-Defect`, `T-Task` to indicate what type of change this is, or add `Type: [enhancement/defect/task]` to the description and I'll add them for you.

### CODE REVIEW

**Comment by su-ex** [comment]
I guess some people don't like the direct connection to YouTube, so this should be an option, probably somewhere in the "Images, GIFs and videos" section of the preference settings.

Have you considered keeping the horizontal layout (embedded video on the left, title and preview text on the right) to utilize wide screens better, optimally breaking it only once the screen gets narrower in width?

**Comment by ayes-web** [comment]
> I guess some people don't like the direct connection to YouTube

I have thought of that, it only connects to Youtube when you click play which seems enough for me.

> Have you considered keeping the horizontal layout (embedded video on the left, title and preview text on the right) to utilize wide screens better, optimally breaking it only once the screen gets narrower in width?

I have not considered that.

**Comment by ayes-web** [comment]
> > > I guess some people don't like the direct connection to YouTube
> > 
> > 
> > I have thought of that, it only connects to Youtube when you click play which seems enough for me.
> 
> So does this use the server-generated url previews, and not download them from yt directly?
> 

Currently its downloading the thumbnail from Youtube.

> > > Have you considered keeping the horizontal layout (embedded video on the left, title and preview text on the right) to utilize wide screens better, optimally breaking it only once the screen gets narrower in width?
> > 
> > 
> > I have not considered that.
> 
> I personally would prefer if the previews stay the size they are now, since I don't plan on using the inline player and don't want the links to waste too much space. People who do want to use it might prefer it bigger, so I'd expect a setting would still be a good idea.

Don't have anything against it, will add the setting.

**Comment by SpiritCroc** [comment]
> Currently its downloading the thumbnail from Youtube.

I imagine some people wouldn't like that - so in a good "privacy by default" manner, we'd make this opt-in (i.e. disabled by default) ;)

We might need to update the privacy policy for communicating with yt :thinking: 
Maybe we can extend the "Integration Services (Bots and Widgets)" section to mention that.

File: src/settings/handlers/AccountSettingsHandler.ts [hunk 1]
```diff
@@ -82,6 +82,12 @@ export default class AccountSettingsHandler extends MatrixClientBackedSettingsHa
     };
 
     public getValue(settingName: string, roomId: string): any {
+        // Special case for YouTube embed player
+        if (settingName === "youtubeEmbedPlayer") {
+            const content = this.getSettings(YOUTUBE_EMBED_PLAYER_EVENT_TYPE);
+            return content ? content["recent_emoji"] : null;
+        }
+
```
Comment (su-ex) [inline src/settings/handlers/AccountSettingsHandler.ts:90]: What is the special case for?

File: src/settings/handlers/AccountSettingsHandler.ts [hunk 2]
```diff
@@ -154,6 +160,14 @@ export default class AccountSettingsHandler extends MatrixClientBackedSettingsHa
...
 
     public async setValue(settingName: string, roomId: string, newValue: any): Promise<void> {
+        // Special case for YouTube embed player
+        if (settingName === "youtubeEmbedPlayer") {
+            const content = this.getSettings(YOUTUBE_EMBED_PLAYER_EVENT_TYPE) || {};
+            content["recent_emoji"] = newValue;
+            await MatrixClientPeg.get().setAccountData(YOUTUBE_EMBED_PLAYER_EVENT_TYPE, content);
+            return;
+        }
+
```
Comment (su-ex) [inline src/settings/handlers/AccountSettingsHandler.ts:170]: ^^

File: src/settings/handlers/AccountSettingsHandler.ts (hunk 1 above)
Comment (ayes-web) [reply src/settings/handlers/AccountSettingsHandler.ts]: Left over unused code, i removed it.

**Review by su-ex — COMMENTED** [review]
This change looks very weird. What's the reason for it failing. Something tells me this is not the right way to solve it.

File: src/components/views/rooms/LinkPreviewWidget.tsx [hunk 3]
```diff
@@ -125,7 +125,8 @@ export default class LinkPreviewWidget extends React.Component<IProps> {
 
         // Youtube video player embed
         const youtubeRegex = /^https?:\/\/(m[.]|www[.])?(youtube[.]com\/watch[?]v=|youtu[.]be\/)([\w-]+)(\S+)?$/;
-        if (this.props.youtubeEmbedPlayerEnabled && this.props.link.match(youtubeRegex)) {
+
+        if (this.props.youtubeEmbedPlayerEnabled || SettingsStore.getValue("youtubeEmbedPlayer") && this.props.link.match(youtubeRegex)) {
```
Comment (su-ex) [inline src/components/views/rooms/LinkPreviewWidget.tsx:129]: This change looks very weird. What's the reason for it failing. Something tells me this is not the right way to solve it.

Comment (ayes-web) [reply src/components/views/rooms/LinkPreviewWidget.tsx:129]: this.props.youtubeEmbedPlayerEnabled gets set to setting value only when its updated, it is undefined by default.

File: src/components/structures/RoomView.tsx [hunk 4]
```diff
@@ -1237,6 +1256,10 @@ export class RoomView extends React.Component<IRoomProps, IRoomState> {
                 // non-e2ee url previews are stored in legacy event type `org.matrix.room.preview_urls`
                 this.updatePreviewUrlVisibility(room);
             }
+
+            if (type === "im.element.setting.youtube_embed_player") {
```
Comment (su-ex) [inline src/components/structures/RoomView.tsx:1260]: ^^

File: src/components/structures/TimelinePanel.tsx [hunk 5]
```diff
@@ -1622,6 +1623,7 @@ class TimelinePanel extends React.Component<IProps, IState> {
                 enableFlair={SettingsStore.getValue(UIFeature.Flair)}
                 hideThreadedMessages={this.props.hideThreadedMessages}
                 disableGrouping={this.props.disableGrouping}
+                youtubeEmbedPlayerEnabled={this.props.youtubeEmbedPlayerEnabled}
```
Comment (su-ex) [inline src/components/structures/TimelinePanel.tsx:1626]: Move below `userNameColorMode`.

File: src/components/views/rooms/EventTile.tsx [hunk 6]
```diff
@@ -342,6 +342,7 @@ interface IProps {
     // displayed to the current user either because they're
     // the author or they are a moderator
     isSeeingThroughMessageHiddenForModeration?: boolean;
+    youtubeEmbedPlayerEnabled?: boolean;
```
Comment (su-ex) [inline src/components/views/rooms/EventTile.tsx:345]: Move.

**Comment by su-ex** [comment]
Thanks for adjusting, code wise this looks good now I think. 👍
I've tested it yesterday and the YouTube previews are gigantic compared to everything else. Maybe it should respect the image size setting in the appearance settings just like other images and videos do.

**Comment by su-ex** [comment]
I've merged upstream. This has now a tiny conflict, but it should be simple to resolve while rebasing on `sc`.

**Comment by ayes-web** [comment]
> I've merged upstream. This has now a tiny conflict, but it should be simple to resolve while rebasing on `sc`. I don't expect that something broke through this upstream merge, but better check if it still works, please.

Should be fine, saw nothing broken

**Review by su-ex — APPROVED** [review]

**Comment by su-ex** [comment]
I've adjusted this a little. Please check if it works as intended for you as well now.
From my side this is good to merge now. :partying_face:

**Comment by su-ex** [comment]
Oh, I've just noticed there is still a connection to YouTube: https://www.youtube-nocookie.com/api/stats/qoe
This needs to vanish, otherwise we need to disable the option by default.

**Review by su-ex — CHANGES_REQUESTED** [review]
No connection to YouTube by default

**Comment by ayes-web** [comment]
> And I assume you'll accept to publish your changes under the Apache-2.0 License? I'll need to find a way to automatically add some checkbox to PRs for that so it can be ticked off and remove that automation stuff inherited from Element ...

Yeah i will, i don't think i need to do anything special for this?

**Comment by su-ex** [comment]
> > And I assume you'll accept to publish your changes under the Apache-2.0 License? I'll need to find a way to automatically add some checkbox to PRs for that so it can be ticked off and remove that automation stuff inherited from Element ...
> 
> Yeah i will, i don't think i need to do anything special for this?

No, I just wanted to confirm this.

**Review by su-ex — APPROVED** [review]

**Comment by su-ex** [comment]
Ok, I can't reproduce the connection anymore either. Maybe it was a remnant from a previous video play test.

**Comment by su-ex** [comment]
Sorry for the delay!

**Pull request merged**